python/ray/train/v2/tests/test_checkpoint_manager.py

```diff
@@ -166,13 +148,53 @@ async def test_before_init_train_context(tmp_path):
     }
 
     # Assert with a checkpoint
-    latest_checkpoint_result = _create_dummy_training_results(1, storage_context)[0]
-    checkpoint_manager.register_checkpoint(latest_checkpoint_result)
+    latest_checkpoint_result = create_dummy_training_results(1, storage_context)[0]
+    checkpoint_manager.register_checkpoint(latest_checkpoint_result, False)
     assert checkpoint_manager.before_init_train_context(workers) == {
         "checkpoint": [latest_checkpoint_result.checkpoint] * 4,
     }
 
 
+async def test_pending_checkpoint_management(tmp_path):
+    storage_context = StorageContext(
+        storage_path=tmp_path,
+        experiment_dir_name="checkpoint_validation_management_experiment",
+    )
+    checkpoint_config = CheckpointConfig(
+        num_to_keep=1,
+        checkpoint_score_attribute="score",
+        checkpoint_score_order="max",
+    )
+    checkpoint_manager = CheckpointManager(
+        storage_context=storage_context,
+        checkpoint_config=checkpoint_config,
+    )
+    training_results = create_dummy_training_results(
+        num_results=3, storage_context=storage_context
+    )
+
+    # Register pending/final checkpoints and verify their storage
+    checkpoint_manager.register_checkpoint(training_results[0], True)
+    checkpoint_manager.register_checkpoint(training_results[2], False)
+    checkpoint_manager.register_checkpoint(training_results[1], True)
+    assert checkpoint_manager._checkpoint_results == [
+        training_results[0],
+        training_results[1],
+        training_results[2],
+    ]
+
+    # Assert checkpoint state after all tasks are done
+    checkpoint_manager.update_checkpoints_with_metrics(
+        {
+            training_results[0].checkpoint: {"score": 100},
+            training_results[1].checkpoint: {"score": 200},
+        }
+    )
+    assert checkpoint_manager._checkpoint_results == [
+        training_results[1],
+    ]
```

Comment:
I kind of expected the better of the 2 "final" checkpoints to stick around as well.

Otherwise the unknown score always beats a finalized score.

Validations must be finalized anyways for the run to exit, so the end state will not violate the num_to_keep config.